--- weston-8.0.0/libweston/backend-rdp/rdp.c.orig	2019-03-28 02:15:13.000000000 +0100
+++ weston-8.0.0/libweston/backend-rdp/rdp.c	2019-04-22 10:01:34.149025872 +0200
@@ -320,7 +320,8 @@
 	if (!nrects)
 		return;
 
-	marker.frameId++;
+	memset(&marker, 0, sizeof(marker));
+	marker.frameId++; /* FIXME: always 1? */
 	marker.frameAction = SURFACECMD_FRAMEACTION_BEGIN;
 	update->SurfaceFrameMarker(peer->context, &marker);
 
